homeassistant/components/combined_energy/coordinator.py

```diff
@@ -0,0 +1,101 @@
+"""DataUpdateCoordinator for the PVOutput integration."""
+from __future__ import annotations
+
+from abc import abstractmethod
+from datetime import timedelta
+from typing import TypeVar
+
+from combined_energy import CombinedEnergy
+from combined_energy.exceptions import CombinedEnergyAuthError, CombinedEnergyError
+from combined_energy.helpers import ReadingsIterator
+from combined_energy.models import DeviceReadings
+
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryAuthFailed
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    LOG_SESSION_REFRESH_DELAY,
+    LOGGER,
+    READINGS_INCREMENT,
+    READINGS_INITIAL_DELTA,
+    READINGS_UPDATE_DELAY,
+)
+
+_T = TypeVar("_T")
+
+
+class _CombinedEnergyCoordinator(DataUpdateCoordinator[_T]):
+    """Get and update the latest data."""
+
+    def __init__(
+        self, hass: HomeAssistant, api: CombinedEnergy, update_interval: timedelta
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            LOGGER,
+            name=type(self).__name__,
+            update_interval=update_interval,
+        )
+        self.api = api
+
+    @abstractmethod
+    async def _update_data(self) -> _T:
+        """Update data."""
+
+    async def _async_update_data(self) -> _T:
+        """Update data with error handling."""
+        try:
+            self.data = await self._update_data()
+        except CombinedEnergyAuthError as ex:
+            raise ConfigEntryAuthFailed from ex
+        except CombinedEnergyError as ex:
+            raise UpdateFailed("Error updating Combined Energy") from ex
+        return self.data
+
+
+class CombinedEnergyLogSessionCoordinator(_CombinedEnergyCoordinator[None]):
+    """Triggers a log session refresh event keep readings data flowing.
+
+    If this is not done periodically, the log session will expire and
+    readings data stops being returned.
+    """
+
+    def __init__(self, hass: HomeAssistant, api: CombinedEnergy) -> None:
+        """Initialize coordinator."""
+        super().__init__(hass, api, LOG_SESSION_REFRESH_DELAY)
+        self.async_add_listener(self.update_listener)
+
+    @staticmethod
+    def update_listener() -> None:
+        """Log that the session has been restarted."""
+        LOGGER.debug("Log session has been restarted")
+
+    async def _update_data(self) -> None:
+        """Update data."""
+        await self.api.start_log_session()
```

Comment:
We have already covered this in a previous review.  It is required, the linked code calls that end point on the _first_ call, but does not refresh the log session.

Reply:
The library can easily be improved to check on each call, how long it has passed since the last log_session call was made, and make the call if it is required.

From the docs the function should do that exactly:
```python
 async def _check_for_log_session_restart(self) -> None:
        """Check if a start log session message is required.
```

After finding this function in the library and checking the library, the functionality of the start_log_session should be there and you should not use a coordinator in a hacky way (subscribe itself as listener) for it.

> We have already covered this in a previous review.
The previous review is obsolete as I find the function and the coordinator performs the call even when no entities are enabled (The entities normally add a listener to the coordinator) as you add the coordinator itself as a listener. The last one is not allowed

Please move the functionality of this coordinator to the library.